### PR TITLE
4.step, please review

### PR DESCRIPTION
**Pull request by kornisb**

### CODE REVIEW

File: pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    private Logger logger = LoggerFactory.getLogger(OrderFacade.class);
+
+    /**
+     * Saves {@link Order}
+     * @param order the {@link Order} to save
+     */
+    public void saveOrder(Order order) {
+        for (PancakeAmount pancakeAmount : order.getOrderedAmounts()) {
+            if (pancakeAmount.getAmount() != 0) {
+                logger.info(createLogMessage(order.getEmail(), pancakeAmount));
```
Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java]: I suggest to log saved order into one line (aka. log only once).

File: pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        }
+    }
+
+    private String createLogMessage(String email, PancakeAmount pancakeAmount) {
+        return new StringBuilder("'")
```
Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java]: Logger has a very good way of using placeholders please consider to use that way.

Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java]: Please check the result of this method.

File: pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class PancakeAmount {
+    private Pancake type;
+    private int amount;
+
+    public Pancake getType() {
+        return type;
+    }
+
+    public void setType(Pancake type) {
```
Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java]: Do set methods really need? if not please remove them.

File: pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    public int getAmount() {
+        return amount;
+    }
+
+    public void setAmount(int amount) {
+        this.amount = amount;
+    }
+
+    private PancakeAmount(Builder builder) {
```
Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java]: Please move right before the first getter.

File: pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        this.type = builder.type;
+        this.amount = builder.amount;
+    }
+
```
Comment (BotondSoltesz) [inline pancake-service/src/main/java/io/github/pancake/service/pancake/domain/PancakeAmount.java]: I suggest to create a _builder()_ method

    public static Builder builder() {
        return new Builder();
    }

to call builder like: PancakeAmount.builder().with...

File: pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                .build();
+        PancakeAmount amount2 = new PancakeAmount.Builder()
+                .withType(Pancake.COCOA)
+                .withAmount(1)
+                .build();
+        when(mockOrder.getOrderedAmounts()).thenReturn(Arrays.asList(amount1, amount2));
+        // WHEN
+        underTest.saveOrder(mockOrder);
+        // THEN
+        verify(mockLogger, times(1)).info(anyString());
```
Comment (BotondSoltesz) [inline pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java]: _times(1)_ is the default. Therefore using _times(1)_ explicitly can be omitted.

File: pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public void setUp() {
+        initMocks(this);
+        underTest = new OrderFacade();
+        underTest.setLogger(mockLogger);
+    }
+
+    @Test
+    public void testSaveOrderShouldLogOrderWhenInvoked() {
+        // GIVEN
+        PancakeAmount amount1 = new PancakeAmount.Builder()
```
Comment (BotondSoltesz) [inline pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java]: Please move creation of amount to a separate method.

File: pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    @BeforeMethod
+    public void setUp() {
+        initMocks(this);
+        underTest = new OrderFacade();
+        underTest.setLogger(mockLogger);
+    }
+
+    @Test
+    public void testSaveOrderShouldLogOrderWhenInvoked() {
```
Comment (BotondSoltesz) [inline pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java]: As per this name I can't expect any filtering but pancake with zero amount will be filtered.

File: pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        // GIVEN
+        PancakeAmount amount1 = new PancakeAmount.Builder()
+                .withType(Pancake.CINNAMON)
+                .withAmount(0)
+                .build();
+        PancakeAmount amount2 = new PancakeAmount.Builder()
```
Comment (BotondSoltesz) [inline pancake-service/src/test/java/io/github/pancake/service/order/facade/OrderFacadeTest.java]: Please try to use better names e.g. _cocoaAmount_ and cinnamonAmount.

File: pancake-webapp/src/main/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformer.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package io.github.pancake.app.order.view.transform;
+
+import org.springframework.stereotype.Component;
+
+import io.github.pancake.service.pancake.domain.PancakeAmount;
+
+@Component
+public class PancakeAmountTransformer {
+    public PancakeAmount transform(io.github.pancake.app.domain.PancakeAmount pancakeAmount) {
```
Comment (BotondSoltesz) [inline pancake-webapp/src/main/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformer.java]: As we discussed next time please consider to add different names to avoid using fully qualified name.

File: pancake-webapp/src/main/java/io/github/pancake/app/order/view/support/OrderFormModelBuilder.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
+package io.github.pancake.app.order.view.support;
+
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.ModelAttribute;
+
+import io.github.pancake.app.order.view.controller.OrderFormController;
```
Comment (BotondSoltesz) [inline pancake-webapp/src/main/java/io/github/pancake/app/order/view/support/OrderFormModelBuilder.java]: Please don't import dependencies if it is used due some javadoc reason rather use fully qualified name in javadoc.

File: pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+public class OrderFacade {
+    private Logger logger = LoggerFactory.getLogger(OrderFacade.class);
+
+    /**
+     * Saves {@link Order}
+     * @param order the {@link Order} to save
+     */
+    public void saveOrder(Order order) {
+        String orderedAmountsMessage = createOrderedAmountsMessage(order.getOrderedAmounts());
+        if (!orderedAmountsMessage.isEmpty()) {
```
Comment (kicsikrumpli) [inline pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java]: I think it is a better approach to log in every case - an empty message also conveys information

File: pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            logger.info("'{}' ordered: {}", order.getEmail(), orderedAmountsMessage);
+        }
+    }
+
+    private String createOrderedAmountsMessage(List<PancakeAmount> orderedAmounts) {
+        StringBuilder builder = new StringBuilder();
+        for (PancakeAmount pancakeAmount : orderedAmounts) {
```
Comment (kicsikrumpli) [inline pancake-service/src/main/java/io/github/pancake/service/order/facade/OrderFacade.java]: a bit wordy, but works fine... how about trying the same with java8 streams: `orderedAmounts.stream().map(...).collect(...)?`

File: pancake-service/src/main/java/io/github/pancake/service/pancake/facade/PancakeFacade.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        return pancakeFactory.getObject();
+    }
+
+    /**
+     * Facade method to get {@link List} of available amounts
+     * @return {@link List} of available amounts
+     */
+    public List<Integer> getAvailableAmounts() {
+        List<Integer> availableAmounts = new ArrayList<>();
+        for (int i = 0; i <= orderLimit; i++) {
```
Comment (kicsikrumpli) [inline pancake-service/src/main/java/io/github/pancake/service/pancake/facade/PancakeFacade.java]: same with java8 streams, just for the sake of practice? ;)

File: pancake-webapp/src/main/java/io/github/pancake/app/confirmation/view/controller/ConfirmationController.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+ */
+@Controller
+public class ConfirmationController {
+    public static final String REQUEST_MAPPING = "/confirmation.html";
+
+    @RequestMapping(value = REQUEST_MAPPING, method = RequestMethod.GET)
+    public String confirmation(@Valid @ModelAttribute("orderRequest") OrderRequest orderRequest, BindingResult bindingResult) {
+        String result;
+        if (bindingResult.hasErrors()) {
+            result = "redirect:";
```
Comment (kicsikrumpli) [inline pancake-webapp/src/main/java/io/github/pancake/app/confirmation/view/controller/ConfirmationController.java]: extract to constant

File: pancake-webapp/src/main/java/io/github/pancake/app/order/view/controller/OrderPostController.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    public OrderFormModel createOrderModel() {
+        return orderFormModelBuilder.getOrderFormModel();
+    }
+
+    @RequestMapping(value = REQUEST_MAPPING, method = RequestMethod.POST)
+    public String order(@Valid @ModelAttribute("orderRequest") OrderRequest orderRequest,
+            BindingResult bindingResult, RedirectAttributes redirectAttributes) {
+        orderRequestValidator.validate(orderRequest, bindingResult);
+
+        String result;
```
Comment (kicsikrumpli) [inline pancake-webapp/src/main/java/io/github/pancake/app/order/view/controller/OrderPostController.java]: a more descriptive name would be nice

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/controller/OrderPostControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    public void testOrderShouldSaveOrderWithFacadeMethodWhenBindingResultHasNoErrors() {
+        // GIVEN
+        OrderRequest orderRequest = new OrderRequest();
+        when(mockBindingResult.hasErrors()).thenReturn(false);
+        // WHEN
+        String result = underTest.order(orderRequest, mockBindingResult, mockRedirectAttributes);
+        // THEN
+        verify(mockOrderRequestValidator, only()).validate(orderRequest, mockBindingResult);
+        verify(mockRedirectAttributes, only()).addFlashAttribute(ORDER_REQUEST, orderRequest);
+
```
Comment (kicsikrumpli) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/controller/OrderPostControllerTest.java]: no need for separation by empty line - if you feel that the method is too long, break it up into multiple ones - if you are testing/asserting/verifying multiple things, it is worth considering breaking it up into multiple test methods (horribile dictu with the same setup)

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/support/OrderRequestValidatorTest.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        orderRequest.setOrderedAmounts(Arrays.asList(cinnamonAmount, cocoaAmount));
+        when(mockPancakeFacade.getOrderLimit()).thenReturn(2);
+        // WHEN
+        underTest.validate(orderRequest, mockErrors);
+        // THEN
+        verify(mockErrors).rejectValue(ORDERED_AMOUNTS, LIMIT_EXCEEDED, new Object[]{3, 2}, "");
+    }
+
+    @Test
+    public void testValidateShouldNotAddErrorWhenLimitDoNotExceeds() {
```
Comment (kicsikrumpli) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/support/OrderRequestValidatorTest.java]: limit is (not) exceeded
or
does not exceed / exceeds limit

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/support/OrderRequestValidatorTest.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        // GIVEN
+        OrderRequest orderRequest = new OrderRequest();
```
Comment (kicsikrumpli) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/support/OrderRequestValidatorTest.java]: quite a bit of repeating code. pls consider moving common elements to setup

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformerTest.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+    @BeforeMethod
+    public void setUp() {
+        underTest = new PancakeAmountTransformer();
+    }
+
+    @Test
+    public void testTransformShouldReturnTransformedObjectWhenInvoked() {
+        // GIVEN
+        io.github.pancake.app.domain.PancakeAmount pancakeAmount =
```
Comment (kicsikrumpli) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformerTest.java]: why the fully qualified name?

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/controller/OrderPostControllerTest.java (hunk 17 above)
Comment (kornisb) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/controller/OrderPostControllerTest.java]: I feel it is a bit too long, but the feeling that all these stuff belong to the same test method is stronger.
I removed the empty lines instead.

File: pancake-webapp/src/test/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformerTest.java (hunk 20 above)
Comment (kornisb) [inline pancake-webapp/src/test/java/io/github/pancake/app/order/view/transform/PancakeAmountTransformerTest.java]: It is because there is a PancakeAmount class in the service layer and in the webapp layer also. The transformer class uses both of them.
